lib/Conversion/TTIRToLinalg/TTIRToLinalg.cpp

```diff
@@ -0,0 +1,206 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Conversion/TTIRToLinalg/TTIRToLinalg.h"
+
+#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"
+#include "ttmlir/Dialect/TTIR/IR/TTIROps.h"
+
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Types.h"
+#include "mlir/IR/Value.h"
+#include "mlir/IR/ValueRange.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/ErrorHandling.h"
+
+using namespace mlir;
+using namespace mlir::tt;
+
+namespace {
+
+using TensorRanks = SmallVector<int64_t, 2>;
+
+static LogicalResult computeBroadcastedShape(SmallVector<Value, 3> inputs,
+                                             TensorRanks &broadcastedShape) {
+  broadcastedShape.clear();
+
+  // Find the maximum rank of all inputs.
+  int64_t maxRank = 0;
+  for (Value input : inputs) {
+    auto type = dyn_cast<RankedTensorType>(input.getType());
+    if (!type) {
+      return failure();
+    }
+    maxRank = std::max(maxRank, type.getRank());
+  }
+
+  // Initialize broadcastedShape to the correct size, one-filled.
+  broadcastedShape = TensorRanks(maxRank, 1);
+
+  // From right-to-left, replace target dim with any non-1 values we encounter
+  // in inputs, returning failure if we find incompatible ranks.
+  for (Value input : inputs) {
+    auto type = dyn_cast<RankedTensorType>(input.getType());
+    const ArrayRef<int64_t> shape = type.getShape();
+
+    for (int64_t i = 0; i < maxRank; ++i) {
+      size_t rightIdx = maxRank - 1 - i;
+      size_t inputRightIdx = shape.size() - 1 - i;
+
+      int64_t targetDim = broadcastedShape[rightIdx];
+      int64_t inputDim =
+          inputRightIdx < shape.size() ? shape[inputRightIdx] : 1;
+
+      if (targetDim != inputDim && targetDim != 1 && inputDim != 1) {
+        return failure();
+      }
+      broadcastedShape[rightIdx] = std::max(targetDim, inputDim);
+    }
+  }
+  return success();
+}
+
+// Helper func to check which dims need to be broadcast and which need to be
+// collapsed.  Assumes that inputShape is broadcast-able to targetShape.
+static void getDimsToBroadcastAndCollapse(
```

Comment:
This logic is much more complicated than I originally expected, but linalg.broadcastOp asserts something like "input rank + len(dimsToBroadcastAlong) == output rank".  We want to support torch-like broadcasting, where ranks with size 1 are implicitly broadcasted => we need to do a tensor.collapseOp in such cases, which requires hairy logic.

I would appreciate careful review/re-review of this logic, especially this function 🙂